### PR TITLE
Switch to using unicode when parsing the command line on windows

### PR DESCRIPTION
**Pull request by Rageoholic**
This should help out with issue #534. Just using GetCommandLineW and then adding in some fixups to make it compile, return utf8 still, and make the tests pass.

### CODE REVIEW

**Comment by LemonBoy** [comment]
The tests are failing because `mips` is a big-endian machine, the result from `nextCodepoint` needs to be byte-swapped or the tests disabled on everything but Windows.

File: lib/std/unicode.zig [hunk 1]
```diff
@@ -574,6 +574,27 @@ pub fn utf16leToUtf8Alloc(allocator: *mem.Allocator, utf16le: []const u16) ![]u8
     return result.toOwnedSlice();
 }
 
+/// Caller must free returned memory.
+pub fn utf16leToUtf8WithNull(allocator: *mem.Allocator, utf16le: []const u16) ![:0]u8 {
+    var result = std.ArrayList(u8).init(allocator);
```
Comment (LemonBoy) [inline lib/std/unicode.zig:579]:
```suggestion
    var result = std.ArrayList(u8).initCapacity(allocator, utf16le.len);
```

File: lib/std/unicode.zig [hunk 2]
```diff
@@ -574,6 +574,27 @@ pub fn utf16leToUtf8Alloc(allocator: *mem.Allocator, utf16le: []const u16) ![]u8
     return result.toOwnedSlice();
 }
 
+/// Caller must free returned memory.
+pub fn utf16leToUtf8WithNull(allocator: *mem.Allocator, utf16le: []const u16) ![:0]u8 {
```
Comment (LemonBoy) [inline lib/std/unicode.zig:578]:
```suggestion
pub fn utf16leToUtf8AllocZ(allocator: *mem.Allocator, utf16le: []const u16) ![:0]u8 {
```

**Comment by Rageoholic** [comment]
Changes are applied. Looks like the W functions *should* give you back little endian codepoints. MS has AFAIK never released a version of windows that *runs* on big endian so I don't care but doing the fixup should be really inexpensive or free and really who puts argument parsing on the fast path anyways. Also we can just run the tests on any given machine and you'll know if you broke it.

**Comment by LemonBoy** [comment]
> Changes are applied.

Tests are still failing because `skip`/`next` are not using `littleToNative`.

> Looks like the W functions should give you back little endian codepoints.

UTF-16 or UCS-2 ?

> Also we can just run the tests on any given machine and you'll know if you broke it.

That's a great argument, I like it :+1: 

> MIPs

The lowercase `s` is making me extra sad for no reason :(

**Comment by Rageoholic** [comment]
> The lowercase s is making me extra sad for no reason :(

I wrote that early in the morning and now it's making me sad. 
>UTF-16 or UCS-2 ?

It's definitely UTF-16. I crawled through MSDN to check. Frankly though they can't port to a big endian architecture without either breaking the people who assumed little endian architecture (given I just did I can't blame them) and people who actually do the fixup on big endian architectures. Hopefully enough people were like "I'll just use a library" and the library did the right thing so that MS can keep it's word without too much pain.
https://docs.microsoft.com/en-us/windows/win32/intl/using-byte-order-marks

**Comment by daurnimator** [comment]
> It's definitely UTF-16. I crawled through MSDN to check

Where/how? MSDN usually fails to note that when they say UTF-16 they really mean UCS-2.

**Comment by Rageoholic** [comment]
It says utf-16 in the link I posted

-------- Original Message --------

**Comment by andrewrk** [comment]
Thanks!

**Pull request merged**